### ISSUE TITLE
support for HiDPI

### ISSUE THREAD

**Issue by alekna** [TigerVNC/tigervnc#369]
On a high DPI (HiDPI) screen TigerVNC is unusable. Would be it possible to implement scaling factor option in TigerVNC client? Remmina seems to support HiDPI out of the box.

Please find a screenshot below. TigerVNC is on the left, Remmina - on the right. Both are connected to the same VNC server.
![screenshot from 2016-10-17 00-07-49](https://cloud.githubusercontent.com/assets/10272121/19420918/5855a816-93fe-11e6-9e02-0e4d7e1b17f8.png)

**Comment by hildred** [comment]
If you are using Xvnc (or vnc server) and you only connect from high resolution displays, the -dpi option to Xvnc may be useful.

**Comment by alekna** [comment]
I tried running it with `-dpi 192`. not much of an improvement.
![screenshot from 2016-10-19 02-10-22](https://cloud.githubusercontent.com/assets/10272121/19499856/67423bba-95a1-11e6-97af-2f08ae2ef5ce.png)

**Comment by hildred** [comment]
try larger numbers like 500 or 1000, also make sure you have a scalable fixed width font.

**Comment by alekna** [comment]
It looks really bad with higher `-dpi`. In my opinion scaling should be done on the client side. It should be font and application agnostic. Just like Reminna does that.
![screenshot from 2016-10-24 22-28-42](https://cloud.githubusercontent.com/assets/10272121/19660708/8d4bd3e2-9a39-11e6-9b13-3a7cec6f23c2.png)

**Comment by zeroimpl** [comment]
FYI I got good results by setting the Linux desktop to a high resolution (3840x1960) and configured for DPI of 192, then just using the **TigerVNC Java client** to scale locally by 50%.

**Comment by imperative** [comment]
In addition to the scaling of the remote view, there seem to be issues with the interface of the C++ client. 

Does it have any interface scaling at all? It seems to use the library FLTK, which does support scaling, and should even read an environment variable FLTK_SCALING_FACTOR. However in practice, changing it does not seem to have any effect. Why is that? Doesn't the client use the scaling features provided by FLTK? Does it override the environment variable setting with some hard-coded value?

**Comment by CendioOssman** [comment]
No, there is no interface scaling at all. That FLTK setting is unfortunately only for the next version of FLTK. We are still targeting the stable 1.3.x branch of FLTK.

**Comment by johann-petrak** [comment]
I was initially happy to find that the TigerVNC Java client has a "ScalingFactor" parameter that even supports "auto" but it does not seem to anything at all. 

The inability to properly scale is a huge problem given that most laptops these days use hidpi displays and that VNC is frequently used to access a server from a laptop. None of the Linux VNC clients seem to do this properly so this has made me switch to Windows for this task where proper scaling works much better.

**Comment by terrancewong** [comment]
Thanks, but my issue is not about "contents' scale", it's about scale of UI of client.

**Comment by timkgh** [comment]
+1 for fixing this on Linux. I use both HiDPI and normal DPI clients against the same server. The TigerVNC server is set to a normal `geometry=1680x1005` which looks fine on the clients with normal DPI screens and also on macOS where the TigerVNC viewer does 2x scaling (a bit blurry but usable). However, the Linux viewer running on a HiDPI laptop is tiny, unusable.
Is there a way to have the Linux viewer scale 2x? I don't want to increase the resolution of the server and scale everything on the server because then the normal clients will not be usable, I believe the correct solution here is for the client to do the scaling, just like Remmina and Vinagre can.

**Comment by quyleanh** [comment]
I have work around for scaling text. However, the problem still remains with UI of software/app running on Linux server.

I use 4k 27-inch display to remote to Linux server. The text, icon and all other elements are so small. I use this command to scale:
`gsettings set org.gnome.desktop.interface text-scaling-factor 1.5`

The problem of text is solved. But the UI of software I use is not. They are still small.

Looking forward to seeing solution.

**Comment by schneis** [comment]
Just came across this issue as I am experiencing the same issue - the UI to connect and remote session are too small to be unusable on my HiDPI laptop.   small.  Having the vncviewer UI be able to scale correctly based on local settings is really needed.

**Comment by BenMcLean** [comment]
This issue also affects me both on Windows and Linux.
On Linux, the GUI is just tiny so I need to use the on screen magnifier to be able to read it at all.
On Windows, I am able to work around the problem using a combination of Windows Magnifier and the "Override high DPI scaling behavior. Scaling performed by: Application" option. Without setting this option, the remote session seems to be zoomed in at 150% when I run DPI scaling at 150% on my Windows PC. (both client and target machine are 1080p)

**Comment by pcrow** [comment]
I used ssvnc for my client because of this issue, and it works great.  It has a "-scale" command-line option, and if my local DPI is over 100, I set it to my DPI/100.  Unfortunately, ssvnc hasn't been updated in ten years, and it may not work with current openssl versions, so I'm looking for new options.  I connect to lots of different servers running various operating systems, and I switch between 4K and 1080p monitors routinely as I move my laptop around.  (I run Linux on my laptop.)

**Comment by prasskaly** [comment]
@CendioOssman Any update on this ? 
It would be great if we could have a sharp VNC session with native resolution choices on Mac. The session is still blurry with available display resolutions and only way that make VNC sharp is by scaling the display using apps such as displayMenu. But, that makes all the other apps blurry.
